### PR TITLE
[pulsar-client-cpp] Fix build failure when CMAKE_BUILD_TYPE is not specified

### PR DESCRIPTION
**Pull request by BewareMyPower**
- [x] Changes comply with the [maintainer guide](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/contributing/maintainer-guide.md)
- [x] SHA512s are updated for each updated download
- [ ] The "supports" clause reflects platforms that may be fixed by this new version
- [ ] Any fixed [CI baseline](https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt) entries are removed from that file.
- [ ] Any patches that are no longer applied are deleted from the port's directory.
- [x] The version database is fixed by rerunning `./vcpkg x-add-version --all` and committing the result.
- [x] Only one version is added to each modified port's versions file.

This patch mainly fixes the error that when `CMAKE_BUILD_TYPE` is not specified with a single-configuration generator, the build will fail because the imported location cannot be found.

For example, in `<cmake-build-dir>/CMakeFiles/<target.dir>/build.cmake`, there could be a line like:

```
<target>: unofficial::pulsar::pulsar-NOTFOUND
```

This patch fixes the error above by making `unofficial::pulsar::pulsar` links debug libraries by default.

In addition, fix the error that Snappy cannot be found on Linux and the logic to find libraries of the pulsar library.

### CODE REVIEW

**Comment by BewareMyPower** [comment]
I verified the patch with the following triplets in GitHub actions, see https://github.com/BewareMyPower/vcpkg/pull/2.
- x64-windows
- x64-windows-static
- x64-linux
- x64-linux-dynamic
- x64-osx
- x64-osx-dynamic

And the `arm64-osx` triplet is verified locally on my m1 laptop.

File: ports/pulsar-client-cpp/unofficial-pulsar-config.cmake [hunk 1]
```diff
@@ -61,6 +56,15 @@ if (NOT TARGET unofficial::pulsar::pulsar)
...
         endif ()
+        if (NOT MSVC)
+            if (CMAKE_BUILD_TYPE STREQUAL "Debug")
+                set_target_properties(unofficial::pulsar::pulsar PROPERTIES
+                    IMPORTED_LOCATION "${_pulsar_library_debug}")
+            else ()
+                set_target_properties(unofficial::pulsar::pulsar PROPERTIES
+                    IMPORTED_LOCATION "${_pulsar_library_release}")
+            endif ()
+        endif ()
```
Comment (Neumann-A) [inline ports/pulsar-client-cpp/unofficial-pulsar-config.cmake:67]:
```suggestion
        if (NOT MSVC)
            if (CMAKE_BUILD_TYPE STREQUAL "Debug")
                set_target_properties(unofficial::pulsar::pulsar PROPERTIES
                    IMPORTED_LOCATION "${_pulsar_library_debug}")
            else ()
                set_target_properties(unofficial::pulsar::pulsar PROPERTIES
                    IMPORTED_LOCATION "${_pulsar_library_release}")
            endif ()
        endif ()
```

This does not work for multi config generators. The way to do this by setting `IMPORTED_LOCATION_<CONFIG>`

Comment (BewareMyPower) [reply ports/pulsar-client-cpp/unofficial-pulsar-config.cmake:67]: I supported multi-config generators here: https://github.com/microsoft/vcpkg/blob/c9140a3b500812ad3206317885860d9553b93f13/ports/pulsar-client-cpp/unofficial-pulsar-config.cmake#L57-L61

With MSVC generator, setting `IMPORTED_LOCATION_RELEASE` and `IMPORTED_LOCATION_DEBUG` is enough. However, with Unix Makefiles, if `IMPORTED_LOCATION` is not set, there will be the following error when running `cmake --build`

```
*** target pattern contains no `%'.  Stop.
```

**Comment by BewareMyPower** [comment]
For Unix Makefiles, when I ran `cmake -B build` and then ran `cd build && make VERBOSE=1`, I found the following static libraries were linked:

- vcpkg_installed/arm64-osx/lib/libpulsar.a
- vcpkg_installed/arm64-osx/debug/lib/libprotobufd.a
- vcpkg_installed/arm64-osx/debug/lib/libcurl-d.a
- vcpkg_installed/arm64-osx/debug/lib/libssl.a
- vcpkg_installed/arm64-osx/debug/lib/libcrypto.a
- vcpkg_installed/arm64-osx/lib/libz.a
- vcpkg_installed/arm64-osx/debug/lib/libzstd.a
- vcpkg_installed/arm64-osx/debug/lib/libsnappy.a 

So I think the debug version of `libpulsar.a` should be linked with the default config. What confused me is that `lib/libz.a` but not `debug/lib/libz.a` is linked .

File: ports/pulsar-client-cpp/unofficial-pulsar-config.cmake (hunk 1 above)
Comment (Neumann-A) [reply ports/pulsar-client-cpp/unofficial-pulsar-config.cmake:67]: You need to specify CMAKE_BUILD_TYPE  since you are using a single config generator

**Comment by BewareMyPower** [comment]
Just found `FindZLIB.cmake` sets `ZLIB_LIBRARY` via `select_library_configurations`, which chooses the release library by default.

```cmake
    if( ${basename}_LIBRARY_DEBUG AND ${basename}_LIBRARY_RELEASE AND
           NOT ${basename}_LIBRARY_DEBUG STREQUAL ${basename}_LIBRARY_RELEASE AND
           ( _isMultiConfig OR CMAKE_BUILD_TYPE ) )
        # ...
    elseif( ${basename}_LIBRARY_RELEASE )
        set( ${basename}_LIBRARY ${${basename}_LIBRARY_RELEASE} )
    elseif( ${basename}_LIBRARY_DEBUG )
        set( ${basename}_LIBRARY ${${basename}_LIBRARY_DEBUG} )
    else()
        set( ${basename}_LIBRARY "${basename}_LIBRARY-NOTFOUND")
    endif()
```

Also, after learning `FindZLIB.cmake`, it seems that `target pattern contains no '%'` was not caused by the `IMPORTED_LOCATION` not set. I will recheck the issue.

File: ports/pulsar-client-cpp/unofficial-pulsar-config.cmake (hunk 1 above)
Comment (BewareMyPower) [reply ports/pulsar-client-cpp/unofficial-pulsar-config.cmake:67]: I figured out the reason. It seems to be caused by the lack of `IMPORTED_CONFIGURATIONS`. After I modified the code above to:

```cmake
        add_library(unofficial::pulsar::pulsar UNKNOWN IMPORTED)
        set_target_properties(unofficial::pulsar::pulsar PROPERTIES
            IMPORTED_CONFIGURATIONS "RELEASE"
            IMPORTED_LOCATION_RELEASE "${_pulsar_library_release}")
        if (_pulsar_library_debug)
            set_target_properties(unofficial::pulsar::pulsar PROPERTIES
                IMPORTED_CONFIGURATIONS "RELEASE;DEBUG"
                IMPORTED_LOCATION_DEBUG "${_pulsar_library_debug}")
            unset(_pulsar_library_debug CACHE)
        endif ()
```

It also worked without specifying `CMAKE_BUILD_TYPE`.

**Comment by BewareMyPower** [comment]
@Neumann-A I updated the PR description and the cmake code, PTAL again. Now with Makefile as the generator, all debug libraries will be linked when `CMAKE_BUILD_TYPE` is not specified. See the verbose output on my local env:

> /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++ -g -arch arm64 -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX13.3.sdk -Wl,-search_paths_first -Wl,-headerpad_max_install_names CMakeFiles/HelloWorld.dir/helloworld.cpp.o -o HelloWorld  vcpkg_installed/arm64-osx/debug/lib/libpulsar.a vcpkg_installed/arm64-osx/debug/lib/libprotobufd.a vcpkg_installed/arm64-osx/debug/lib/libcurl-d.a vcpkg_installed/arm64-osx/debug/lib/libssl.a vcpkg_installed/arm64-osx/debug/lib/libcrypto.a vcpkg_installed/arm64-osx/debug/lib/libz.a -framework SystemConfiguration -framework CoreFoundation -framework CoreServices vcpkg_installed/arm64-osx/debug/lib/libzstd.a vcpkg_installed/arm64-osx/debug/lib/libsnappy.a

**Comment by JonLiu1993** [comment]
Tested usage successfully by `pulsar-client-cpp:x64-windows`:
````
pulsar-client-cpp provides CMake targets:

  find_package(unofficial-pulsar CONFIG REQUIRED)
  target_link_libraries(main PRIVATE unofficial::pulsar::pulsar)
````

**Pull request merged**